File: src/main/java/de/maggiwuerze/xdccloader/service/IrcBotService.java
package de.maggiwuerze.xdccloader.service;

import de.maggiwuerze.xdccloader.irc.IrcBot;
import de.maggiwuerze.xdccloader.irc.IrcEventListener;
import de.maggiwuerze.xdccloader.model.download.Download;
import de.maggiwuerze.xdccloader.model.entity.Bot;
import de.maggiwuerze.xdccloader.model.entity.Channel;
import de.maggiwuerze.xdccloader.model.entity.Server;
import de.maggiwuerze.xdccloader.model.forms.TargetBotForm;
import de.maggiwuerze.xdccloader.persistence.TargetBotRepository;
import de.maggiwuerze.xdccloader.util.IpHelper;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.RandomStringUtils;
import org.pircbotx.Configuration;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
@Slf4j
public class IrcBotService {

	private final IrcEventListener ircEventListener;
	private final Map<Bot, IrcBot> activeBots = new HashMap<>();

	public IrcBot getIrcBotForDownload(Download download) {
		return Optional.ofNullable(activeBots.get(download.getBot())).orElseGet(() -> getNewBot(download));
	}

	public IrcBot getNewBot(Download download) {

		String username = RandomStringUtils.randomAlphabetic(1) + RandomStringUtils.random(7, true, true);
		Bot targetBot = download.getBot();

		Configuration configuration = new Configuration.Builder()
			.setName(username) //Set the nick of the bot. CHANGE IN YOUR CODE
			.addServer(targetBot.getServer().getServerUrl()) //Join the freenode network
			.addAutoJoinChannel(targetBot.getChannel().getName()) //Join the official #pircbotx channel
			.setAutoReconnect(true)
			.setAutoReconnectAttempts(5)
			.setAutoNickChange(true) //Automatically change nick when the current one is in use
			.addListener(ircEventListener) //Add our listener that will be called on Events
			.setDccPublicAddress(IpHelper.getPublicIp())
			.setDccPorts(List.of(10000))
			.buildConfiguration();
		IrcBot bot = new IrcBot(configuration, download.getId());
		activeBots.put(targetBot, bot);

		return bot;
	}
}

File: src/main/kotlin/de/maggiwuerze/xdccwebloader/service/IrcBotService.kt
package de.maggiwuerze.xdccwebloader.service

import de.maggiwuerze.xdccwebloader.irc.IrcBot
import de.maggiwuerze.xdccwebloader.irc.IrcEventListener
import de.maggiwuerze.xdccwebloader.model.download.Download
import de.maggiwuerze.xdccwebloader.persistence.entity.Bot
import de.maggiwuerze.xdccwebloader.util.IpHelper
import org.pircbotx.Configuration
import org.springframework.stereotype.Service
import kotlin.random.Random

@Service
class IrcBotService(private val ircEventListener: IrcEventListener) {
    private val activeBots: MutableMap<Bot, IrcBot> = HashMap()

    fun getIrcBotForDownload(download: Download): IrcBot {
        return activeBots[download.bot] ?: getNewBot(download)
    }

    fun getNewBot(download: Download): IrcBot {
        val username = "${randomString()}_${Random.nextInt(10)}"
        val targetBot: Bot = download.bot

        val configuration: Configuration = Configuration.Builder()
            .setName(username) //Set the nick of the bot. CHANGE IN YOUR CODE
            .setOnJoinWhoEnabled(false)
            .addServer(targetBot.server.serverUrl) //Join the freenode network
            .addAutoJoinChannel(targetBot.channel.name) //Join the official #pircbotx channel
            .setAutoReconnect(true)
            .setAutoReconnectAttempts(5)
            .setAutoNickChange(true) //Automatically change nick when the current one is in use
            .addListener(ircEventListener) //Add our listener that will be called on Events
            .setDccPublicAddress(IpHelper.publicIp)
            .buildConfiguration()

        IrcBot(configuration, download.id).let { bot ->
            activeBots[targetBot] = bot
            return bot
        }

    }

    val charPool: List<Char> = ('a'..'z') + ('A'..'Z')
    fun randomString() = (1..7)
        .map { Random.nextInt(0, charPool.size).let { charPool[it] } }
        .joinToString("")
}


